Added angular quantities printing in latex and pretty

Subhash-Saurabh opened the pull request:
Added a function `_print_Quantity` that allows latex printing of degree and radian.
```
>>> latex(90*degree)
r"90 ^\circ"
```
Also made changes so that quantities print at the end of a expression.
```
>>> expr = 2*kg*x*meter**2
>>> latex(expr, mul_symbol='dot')
2 \cdot x \cdot kilogram \cdot meter^{2}
```
#### References to other Issues or PRs
Fixes #12874

Subhash-Saurabh committed:
Added degree printing in latex

gxyd commented on sympy/printing/latex.py:
```diff
@@ -2058,6 +2058,10 @@ def _print_primeomega(self, expr, exp=None):
+    def _print_Quantity(self, expr):
+        if str(expr) == 'degree':
+            return r"^\circ"
+        return r"%s" %expr
```
You should add test for this case as well.

gxyd replied:
Wait.

gxyd commented on sympy/printing/latex.py:
```diff
@@ -2058,6 +2058,10 @@ def _print_primeomega(self, expr, exp=None):
+    def _print_Quantity(self, expr):
+        if str(expr) == 'degree':
+            return r"^\circ"
+        return r"%s" %expr
```
No need for a test case for this line.
Just add a space between `%` and `expr`.

gxyd reviewed (commented):
Also improve the pretty printing may be?

Subhash-Saurabh committed:
Added space between % and expr

gxyd commented:
You can also add latex and pretty printing for other angle units (radian, steradian, ...) as well. If you don't intend to work on that in this PR, I can open an issue to keep track of that.

gxyd commented:
The wiki page for Radian mentions it using both the symbols `rad` and `^c` being in use, I'm not sure of the benefits of one compared to the other. Opinion @jksuom ? I would be more inclined towards using simply `rad`, and not `^c`, since the earlier will create no confusion.

Subhash-Saurabh commented:
I am working to add degree for `prettyprinting` . I will make additional commits as soon as possible.

asmeurer commented on sympy/printing/latex.py:
```diff
@@ -2058,6 +2058,10 @@ def _print_primeomega(self, expr, exp=None):
         return r'\Omega\left(%s\right)' % self._print(expr.args[0])
 
+    def _print_Quantity(self, expr):
+        if str(expr) == 'degree':
```
Use self.name. Don't rely on the str printer here.

Subhash-Saurabh replied:
`Latexprinter` class has no attribute `name` . So using `self.name` throws `AttributeError` . Please point out if I am getting something wrong.

ylemkimon replied:
I think he meant `expr.name`.

Subhash-Saurabh replied:
Type of `expr.name` is Symbol. 
Therefore we have to add check like 
```
if expr.name == Symbol('degree'):
    return r"^\circ"
``` 
Is this format alright or is there something that I am missing.

ylemkimon replied:
You can do that or you can do `expr.name.name`.

Subhash-Saurabh committed:
Added degree printing for prettyprint

Subhash-Saurabh committed:
Replaced str(expr) with expr.name.name

Subhash-Saurabh committed:
Solved incorrect latex and pprint output

When printing expr such as Symbol*degree the printer gave wrong output
that was degree*Symbol .

ylemkimon commented:
Is this ready to merge or are you going to add other units, e.g., radian?

Subhash-Saurabh commented:
I am waiting for someone to approve the changes . Once this gets approved I will add other units in the similar manner.

ylemkimon commented on sympy/printing/latex.py:
```diff
@@ -399,6 +402,12 @@ def convert(expr):
                 else:
                     args = expr.args
```
`expr.args` is a tuple, which is immutable and doesn't support `del`. You have to convert it to a `list`.

Subhash-Saurabh committed:
Added radian printing for latex and pretty

Subhash-Saurabh commented:
@gxyd can you please review this PR.
For steradian printing I think the empty printer is fine . What do you suggest ?

ylemkimon commented on sympy/printing/latex.py:
```diff
@@ -397,7 +400,13 @@ def convert(expr):
+
+                # If angular quantity is present append them at the back
+                for i in range(len(args)):
+                    if str(args[i]) in angular_quantity:
```
Don't rely on the `str` printer here, too.

Subhash-Saurabh replied:
If we use `args[i].name` instead of `str(args[i])` we get error when `args[i]` is Integer,Pow etc. Is there any other way so that we don't have to use the `str` printer and we can make this check.

Subhash-Saurabh committed:
Removed use of str printer in latex and pretty

Subhash-Saurabh commented:
@ylemkimon I have made the changes you suggested .Can you please review it now.

ylemkimon commented:
I think this can be applied to all units.

Subhash-Saurabh commented:
Current behaviour of latex and pprint
For steradian
```
>>> expr = 4*pi*steradian
>>> pprint(expr)
4⋅π⋅steradian
>>> latex(expr)
4 \\pi steradian
```
The [wiki](https://en.wikipedia.org/wiki/Steradian) mentions the above expr to be represented as `4π sr`.
This is the case for other units as well. So I think the current behaviour of `pprint` is fine.
@ylemkimon @gxyd can you please specifically mention what changes do you want in printing of other units as I am unable to figure it out.

Subhash-Saurabh commented:
@ylemkimon I think for pretty printing of other units I will open a new PR also addressing the issue #13269 . So this PR can be merged maybe.

ylemkimon commented on sympy/printing/latex.py:
```diff
@@ -397,7 +401,14 @@ def convert(expr):
+                # If angular quantity is present append them at the back
+                for i in range(len(args)):
+                    if isinstance(args[i], Quantity):
+                        if args[i].name.name in angular_quantity:
```
I was thinking that printing the units at the end can be applied to all units. But if you think it's out of scope of this PR, will merge this PR.

Subhash-Saurabh replied:
I got your point now . I will add other quantities in this PR only.

Subhash-Saurabh committed:
Added pretty printing of other quantities

Subhash-Saurabh committed:
Added latex printing for other quantities

Subhash-Saurabh commented:
@ylemkimon I have added for other units printing as well. Can you please take a look?

ylemkimon commented on sympy/printing/latex.py:
```diff
@@ -397,7 +398,21 @@ def convert(expr):
+                    for i in range(len(args)-1, -1, -1):
+                        if quantities_index != [] and i == quantities_index[-1]:
+                            del quantities_index[-1]
+                            del args[i]
```
You can do `sorted(args, key=lambda x: isinstance(x, Quantity) or (isinstance(x, Pow) and isinstance(x.base, Quantity)))`.

Subhash-Saurabh committed:
Used lambda to sort the args so that quantities remain at the end of list.

ylemkimon commented on sympy/printing/latex.py:
```diff
@@ -2058,6 +2063,12 @@ def _print_primeomega(self, expr, exp=None):
+        if expr.name.name == 'degree':
+            return r"^\circ"
+        elif expr.name.name == 'radian':
+            return r"^rad"
```
Should `rad` be raised?

asmeurer replied:
If it is it needs to be `^{rad}`

Subhash-Saurabh commented:
@ylemkimon The Wikipedia page of radian mentions radian to be expressed in the following form. If you want any changes kindly do suggest.

ylemkimon commented:
But I'm not sure it's commonly used.

Subhash-Saurabh commented:
@ylemkimon Then should we revert back to emptyprinter for radian printing or do you suggest some other changes.

Subhash-Saurabh committed:
Reverted radian printing to earlier form

Subhash-Saurabh commented:
@ylemkimon I have made the suggested changes . Can you please review this PR.

Subhash-Saurabh commented:
@ylemkimon No one is suggesting any change on this PR so I think it can be merged .
